kubespawner/objects.py

```diff
@@ -75,63 +98,86 @@ def make_pod_spec(
         to have access to. String ins loat/int since common suffixes
         are allowed
     """
-    pod_security_context = {}
-    if run_as_uid is not None:
-        pod_security_context['runAsUser'] = int(run_as_uid)
+    api_client = ApiClient()
+
+    pod = V1Pod()
+    pod.kind = "Pod"
+    pod.api_version = "v1"
+
+    pod.metadata = V1ObjectMeta()
+    pod.metadata.name = name
+    pod.metadata.labels = labels.copy()
+    
+    pod.spec = V1PodSpec()
+
+    security_context = V1PodSecurityContext()
     if fs_gid is not None:
-        pod_security_context['fsGroup'] = int(fs_gid)
-    image_secret = []
+        security_context.fs_group = int(fs_gid)
+    if run_as_uid is not None:
+        security_context.run_as_user = int(run_as_uid)
+    pod.spec.security_context = security_context
+
     if image_pull_secret is not None:
-        image_secret = [{"name": image_pull_secret}]
-    return {
-        'apiVersion': 'v1',
-        'kind': 'Pod',
-        'metadata': {
-            'name': name,
-            'labels': labels,
-        },
-        'spec': {
-            'securityContext': pod_security_context,
-            "imagePullSecrets": image_secret,
-            'containers': [
-                {
-                    'name': 'notebook',
-                    'image': image_spec,
-                    'args': cmd,
-                    'imagePullPolicy': image_pull_policy,
-                    'ports': [{
-                        'containerPort': port,
-                    }],
-                    'resources': {
-                        'requests': {
-                            # If these are None, it's ok. the k8s API
-                            # seems to interpret that as 'no limit',
-                            # which is what we want.
-                            'memory': mem_guarantee,
-                            'cpu': cpu_guarantee,
-                        },
-                        'limits': {
-                            'memory': mem_limit,
-                            'cpu': cpu_limit,
-                        }
-                    },
-                    'env': [
-                        {'name': k, 'value': v}
-                        for k, v in env.items()
-                    ],
-                    'volumeMounts': volume_mounts
-                }
-            ],
-            'volumes': volumes
-        }
-    }
+        pod.spec.image_pull_secrets = []
+        image_secret = V1LocalObjectReference()
+        image_secret.name = image_pull_secret
+        pod.spec.image_pull_secrets.append(image_secret)
+
+    pod.spec.containers = []
+
+    notebook_container = V1Container()
```

Comment:
We could also construct notebook_container this way (passing params to __init__) since that supports it too (https://github.com/kubernetes-incubator/client-python/blob/84bf6cc9dc206cb3314a5b4d2fa1a9c2fa2c5137/kubernetes/client/models/v1_container.py). I'm not super attached to it here, though - because there's a lot more params.